### PR TITLE
[SPARK-16526][SQL] Benchmarking Performance for Fast HashMap Implementations

### PR DESCRIPTION
**Pull request by ooq**
## What changes were proposed in this pull request?

The 3rd PR in its series to resolve SPARK-16523.

This patch adds benchmark tests for vectorized hashmap vs. row-based hashmap (along with results in the comments). Those tests are ignored by default as they take long to run.
## How was this patch tested?

This patch are mostly tests itself.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/AggregateBenchmark.scala [hunk 1]
```diff
@@ -576,4 +576,605 @@ class AggregateBenchmark extends BenchmarkBase {
...
+
+    var timeStart: Long = 0L
+    var timeEnd: Long = 0L
+    var nsPerRow: Long = 0L
+    var i = 0
+    sparkSession.conf.set("spark.sql.codegen.wholeStage", "true")
+    sparkSession.conf.set("spark.sql.codegen.aggregate.map.columns.max", "30")
+
+    // scalastyle:off
+    println(Benchmark.getJVMOSInfo())
```
Comment (sameeragarwal) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/AggregateBenchmark.scala]: nit: to minimize duplication, maybe create a small utility function that can then be reused in all test cases.

**Comment by sameeragarwal** [comment]
Just a minor nit. LGTM.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/AggregateBenchmark.scala [hunk 2]
```diff
@@ -576,4 +576,605 @@ class AggregateBenchmark extends BenchmarkBase {
...
+                       6                   61                   29                   23
+                       7                   69                   36                   25
+                       8                   78                   37                   27
+                       9                   88                   43                   30
+                      10                   92                   45                   33
+    Unit: ns/row
+    */
+  }
+
+  ignore("varying key fields, varying value field, 16 linear distinct keys") {
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/AggregateBenchmark.scala]: The performance difference between column-based and row-based are cache locality, could you increase the number of distinct keys to make sure that not all the keys/values are fit in L1 cache? for example, 4k. We could also increase that to 64k in first two cases (single key, single value).

**Comment by ooq** [comment]
@davies Added some test results with larger number of distinct keys.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/AggregateBenchmark.scala [hunk 3]
```diff
@@ -1078,6 +1078,146 @@ class AggregateBenchmark extends BenchmarkBase {
...
+                   16384                   72                   48                   42
+                   32768                   77                   54                   47
+                   65536                   96                   75                   61
+                  131072                  115                  119                  130
+                  262144                  137                  162                  185
+    Unit: ns/row
+    */
+  }
+
+  ignore("single key field, single value field, varying linear distinct keys") {
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/AggregateBenchmark.scala]: Should we access them in random way?

**Comment by HyukjinKwon** [comment]
(gentle ping @ooq)

**Pull request closed without merging**